File: JavaPractice/src/EncodeDecodeTinyURL.java

import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.regex.*;

public class EncodeDecodeTinyURL {

	static Map<String, String> hm = new HashMap<>();
	static String key = null;
	static String returnURL = "http://tinyURL.com/";
	
	static void tinyURLEncodeDecode(String inputString) {

		if(inputString == null || inputString.isEmpty() || inputString.length() == 0 || !validURL(inputString))
			throw new IllegalArgumentException("Pass valid input");
		
		key = getRandomValue();
		String shortURL;
		System.out.println(shortURL = encodeURL(inputString));
		System.out.println(decodeURL(shortURL));
		
	}
	
	static String encodeURL(String longURL) {
		while(hm.containsKey(key)) {
			key = getRandomValue();
		}
		hm.put(key, longURL);
		return returnURL + key;
	}
	
	static String decodeURL(String shortURL) {
		return hm.get(shortURL.replace(returnURL, ""));
	}
	
	static String getRandomValue() {
		
		String alphabet = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
		Random rm = new Random();
		StringBuilder sb = new StringBuilder();
		for(int i=0; i<10; i++) {
			sb.append(alphabet.charAt(rm.nextInt(62)));
		}
		return sb.toString();
	}
	static boolean validURL(String inputString) {
		String regex1 = "^(https://|http://|www.)[-a-zA-Z0-9+/%?_|:.]*[-a-zA-Z0-9](\\.[a-z]{2,3})$";
		try {
			Pattern p = Pattern.compile(regex1);
			Matcher m = p.matcher(inputString);
			if(m.matches())
				return true;
		}
		catch (Exception e) {
			e.printStackTrace();
		}
		return false;
	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		tinyURLEncodeDecode("https://www.helix.com");
		tinyURLEncodeDecode("https://www.abcdefghijladfafacsdcas.com");
		tinyURLEncodeDecode("www.timeofindiaahmedabadtobombay.com");
		tinyURLEncodeDecode("www.vyavsaividhyapratishthanrajkotaliasvvp.com");
		tinyURLEncodeDecode("https://leetcode.com/problems/design-tinyurl");
		tinyURLEncodeDecode("www.om");
		

	}

}

File: PythonPractice/src/EncodeDecodeTinyURL.py
import random
import re
from typing import Optional, Dict

class EncodeDecodeTinyURL:
    _hm: Dict[str, str] = {}
    _return_url: str = "http://tinyURL.com/"

    @staticmethod
    def tiny_url_encode_decode(input_string: str) -> None:
        if not isinstance(input_string, str) or not input_string or not EncodeDecodeTinyURL._valid_url(input_string):
            raise ValueError("Pass valid input")

        key = EncodeDecodeTinyURL._get_random_value()
        short_url = EncodeDecodeTinyURL._encode_url(input_string, key)
        print(short_url)
        print(EncodeDecodeTinyURL._decode_url(short_url))

    @staticmethod
    def _encode_url(long_url: str, key: str) -> str:
        while key in EncodeDecodeTinyURL._hm:
            key = EncodeDecodeTinyURL._get_random_value()
        EncodeDecodeTinyURL._hm[key] = long_url
        return EncodeDecodeTinyURL._return_url + key

    @staticmethod
    def _decode_url(short_url: str) -> Optional[str]:
        key = short_url.replace(EncodeDecodeTinyURL._return_url, "")
        return EncodeDecodeTinyURL._hm.get(key)

    @staticmethod
    def _get_random_value() -> str:
        alphabet = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
        return ''.join(random.choice(alphabet) for _ in range(10))

    @staticmethod
    def _valid_url(input_string: str) -> bool:
        if not isinstance(input_string, str):
            return False

        regex = r"^(https://|http://|www\.)[-a-zA-Z0-9+/%?_|:.]*[-a-zA-Z0-9](\.[a-z]{2,3})$"
        pattern = re.compile(regex)
        return bool(pattern.match(input_string))

# Example usage
try:
    EncodeDecodeTinyURL.tiny_url_encode_decode("https://www.helix.com")
    EncodeDecodeTinyURL.tiny_url_encode_decode("https://www.abcdefghijladfafacsdcas.com")
    EncodeDecodeTinyURL.tiny_url_encode_decode("www.timeofindiaahmedabadtobombay.com")
    EncodeDecodeTinyURL.tiny_url_encode_decode("www.vyavsaividhyapratishthanrajkotaliasvvp.com")
    EncodeDecodeTinyURL.tiny_url_encode_decode("https://leetcode.com/problems/design-tinyurl")
    EncodeDecodeTinyURL.tiny_url_encode_decode("www.om")
except Exception as e:
    print(f"An error occurred: {e}")
